### ISSUE TITLE
Adding event table to activity streams

### ISSUE THREAD

**Issue by scd** [theCrag/website#1736]
A number of issues are now blocked until an event table (plus association table) is added to the activity stream:

- [x] event date based on timezone estimate of crag
- [x] starting discussions against events - issue #1728
- [x] pagination of streams - issue #1725 
- [x] dateless feed of activity streams - issue #1725 
- [x] area stream
- [ ] remove automated favorite events due to signing up to the app

I don't think this is difficult, but I want to beta release streams without event table so we can collect further feedback.

**Comment by scd** [comment]
Working post to help extend the table structure and get @scd back up to spead on streams. Based on a quick review of code in the working demo.

StreamEvent
  CreateDate: system time when the event was first created
  UpdateDate: system time when the event has last had activity associated with it
  Timezone: based on crag, or account (if no crag).  Used for calculating StreamDate
  StreamType: type of stream (eg tick, update, follow, discuss, favorite, upload, join, customise)
  StreamSouce: account (or other) entity for aggregating stream (eg in the case of discussion message, this would be the root message id)
  StreamPrivacyAccount: if present then use this account for a privacy check (eg ascent). If not present then no privacy check is required (eg public discussions).
  StreamSeed: For streams that are aggregated on something other than the account that submitted the update (eg route message for discussions)
  StreamDate: date based on estimated timezone
  StreamNode: Gym/TLC/World
  Nx: 

StreamAssociation
  StreamEvent
  ActivityLogItem

StreamEvents are aggregated on StreamDate, StreamSource, StreamNode and StreamType.

The StreamDate is the aggregation date based on the estimated timezone of the crag.

For most use cases the StreamSource will be the account that made the update. However there is the opportunity to aggregate on something other then the account id. In particular this applies to discussions, where the actual stream seed is the route message. This means that discussion item from multiple accounts can be grouped in the same stream.

Permissions to see the stream or not come from the StreamPrivacyAccount. If not set then there is no privacy check required.

**Comment by scd** [comment]
More on StreamSource...

What I am trying to do is get a logical model that matches our beta product, in particular the differences between three models which we have identified in the beta product.

1. Normal update associated with a person (eg update route description / log tick), where all matching update types are grouped on a stream event for a particular account and crag.

2. A discussion in a crag forum where the discussion is a stream event in itself. The difference between this model and the one above is that an account can start multiple discussions in the same crag, each of which should be grouped separately and therefore the underlying grouping cannot be the account, but rather the discussion id.

3. Comments associated with an event (equivalent to commenting on an ascent). This is similar to the discussion model in that an account can make multiple comments. In this model, somebody makes multiple comments on in one event then it should have the same source aggregation. From a UI point of view, this model is like a virtual stream event which points to the original event and will bring in the original event.

There are three distinct behaviours all driven by the source aggregation. In model 1, the source aggregation is the account, in model 2 it is the discussion id and in model 3 it is another event.

**Comment by scd** [comment]
More on StreamEventOwner ...

An event should have an owner so when you do a search for events that are 'mine' it is easy to pickup. This is simple for model 1, above. However has implications in model 2 and model 3. I really want to keep the paradigm and implementation simple if possible.

If we make the following assumptions for stream filtering.

 - [ ] For 'mine' discussions the system will filter on discussions you started. It will not necessarily pull in discussions you contributed to.
 - [ ] For discussions of accounts you follow the system will filter on discussions each account started, but not necessarily on discussions that they contributed to.
 - [ ] If somebody makes a comment on one of your events (eg ascent) then this will be filtered in the 'mine' filter as you are the owner of that event.
 - [ ] If you comment on somebody else's event it will not appear on your 'mine' filter.
 - [ ] You will get event comments that are made against events of people you are following.
 - [ ] If somebody you are following makes a comment against an event owned by somebody you are not following then you may not get this event.

I think these restrictions are ok for now because we have an alternate way of following discussions. Once you are involved in a discussion you will get it in your dashboard.

If these assumptions are not ok then we will have to create and join on an 'Event Involved' table (or join on the massive activity table).

BTW, the streams cannot replace our dashboard discussions because streams do not do private discussions, just public discussions. There may be a case to replace dashboard discussions in the fullness of time, but with this next milestone.

**Comment by scd** [comment]
More on StreamEventDate ...

Business rules for creating/updating this:
 - [ ] Use a timezone offset approximation based on geo location of crag
 - [ ] If crag is not geo located use GMT
 - [ ] Note that the process of geolocating a crag could cause some once-off stream date idiosyncrasies. Because they are once off and the nature of the streams is fleeting anyway, this will be acceptable.
 - [ ] For crag discussion events use the last date of last entry. This means a discussion event date may change as further contributions are made.
 - [ ] Similarly for ascent comments the event date will be the date of the last comment. The difference discussion events is that ascent comments is a virtual event, which does not alter the original ascent logging event. In otherwords a users ascent logging event will keep it's date. It is up to the UI to make sure the events are not displayed twice on one page.

**Comment by brendanheywood** [comment]
* A crag should always have an inherited location, if not located then use it's parent and so on. Should always get to a country at least and they should all be located, think there is already inherited location data we can leverage for this.

Also when working out the dates, make the boundary say 3am instead of midnight, we get a lot of edits later at night and that makes more sense to merge them than split them.

Also I think StreamEventOwner should accept multiple accounts. Typical use cases in facebook land are:
* I post something and mention a friend in the post, this is pretty well equivalent to if they posted it themselves from a stream perspective. If I didn't do this but then later mentioned them in a comment on that post it is handled quite differently (in facebook).
* Same as above for photos, if I tag the climber and belayer in a photo, and I took it, should be handled in a stream roughly the same way for all three accounts. Caveats to this may be that we only do this for linked accounts, but I don't really like that because we don't have reciprocal 'friend's like facebook, so the other alternative is to 'block accounts' which negates the linking. Can be done later.
* Probably should be the same for ticking, if I batch tick 10 routes with 3 other people, then add that to their streams. Similar to discussions need to de-dup somehow if the other people also ticked as well.

**Comment by scd** [comment]
>  Also I think StreamEventOwner should accept multiple accounts. Typical use cases in facebook land are:

I was hoping to avoid this, which is why I list the assumptions. However based on your comments I think that we need to have multiple people involved in an event. I have made more comments below because there are some differences between what we are doing and facebook and I want to make sure we are on the same page.

 >    I post something and mention a friend in the post, this is pretty well equivalent to if they posted it themselves from a stream perspective. If I didn't do this but then later mentioned them in a comment on that post it is handled quite differently (in facebook).

People will configure what they want in their streams, so we need to define our stream behavior in terms of 'mine' and 'following' configuration.

I think what you are suggesting here is that if you are 'mentioned' in a public crag discussion then you should see that discussion if you have select the Show discussions 'mine' checkbox.

If we are going down this path for crag discussions I don't think it should matter when you are mentioned. Also if you contribute to the discussion it should turn up on 'mine' stream. For this functionality I will need multiple accounts.
 

>    Same as above for photos, if I tag the climber and belayer in a photo, and I took it, should be handled in a stream roughly the same way for all three accounts. Caveats to this may be that we only do this for linked accounts, but I don't really like that because we don't have reciprocal 'friend's like facebook, so the other alternative is to 'block accounts' which negates the linking. Can be done later.

Same logic. Anywhere where we have mentions this should appear in the 'mine' stream. Once you are mentioned in a discussion/photo/fa/tick/wherever you should be automatically included as involved in the event.

 >    Probably should be the same for ticking, if I batch tick 10 routes with 3 other people, then add that to their streams. Similar to discussions need to de-dup somehow if the other people also ticked as well.

De-duping will happen automatically for this because we will select distinct events. This is part an parcel of the filtering, because events may appear in several of the filter queries (following a person and watching a crag).

In summary I need a second table:

StreamEventInvolved
  StreamEvent
  Account

I think this was a very worthwhile discussion.

**Comment by brendanheywood** [comment]
Yup we are on the same page. This is how github, facebook, twitter, pinterest, etc work so it makes sense to follow the crowd (and I also think it is this way because it's better too)

Facebook does handle the post mention vs a comment mention slightly differently, I @scd posts a photo and tags @cgome in it, and I follow @cgome then I will see it in my stream if I follow @cgome. It is was merely a comment mention then @cgome will be added to the discussion and get notifications, but I won't. This is subtle and I'm not fussed if we deviate from FB is one way is easier than the other.

> De-duping will happen automatically for this because we will select distinct events.

Yes, but what I was referring too is a little different because there are two events, one is @scd ticking 10 routes which mention @cgome, and the other could be @cgome ticking the same routes which mention @scd. If I follow either or both of you I will see both events. I think we just show both as I can't see an easy way to deal with this, and it's probably better to show both anyway.

**Comment by scd** [comment]
> Yes, but what I was referring too is a little different because there are two events, one is @scd ticking 10 routes which mention @cgome, and the other could be @cgome ticking the same routes which mention @scd. If I follow either or both of you I will see both events. I think we just show both as I can't see an easy way to deal with this, and it's probably better to show both anyway

This is definitely two separate events in our model. One is 'Simon ticks 10 routes in Arapiles' and the other is "Campbell ticks 6 routes in Arapiles". It just happens we were both involved in each others ticks.

**Comment by scd** [comment]
FYI, today I did a drop-in replace of our new stream table in the stream facet. Our idea worked perfectly first go. Now we are in a position for me to fix up all these dependent issues. Even though there was only about an hours work to replace the aggregation with the table, there was quite a bit of work designing the table and doing the data migration. 

The new table uses crag time with day switch over being 3am. 

Unless there is any hidden gotchas the new table also takes into account discussions on events, ascents, photos, etc. I have been thinking about this and we definitely need to be able to discuss individual objects as well as object groupings on an event. Either way discussions about events/objects will be connected to the event itself.

This is going to be really really cool.

**Comment by brendanheywood** [comment]
:tada: :fireworks: :smile:
